### PR TITLE
Add support for Cloud's new rate limit headers 

### LINKED ISSUE #108: Controlplane: Handle RateLimit header field from controlplane
Cloud's control plane implements RateLimit header fields https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-ratelimit-headers to advertise their rate limiting policy.

The provider should handle these and wait accordingly, for example:
```
< ratelimit: limit=500, remaining=0, reset=5
< ratelimit-policy: 500;w=60;policy="account"
< ratelimit-policy: 50000;w=60;policy="organization"
```

limit = 500, remaining = 0 (limited already), reset in 5s

In this case, the provider should wait before sending the next request and avoid failing the resource creation if the headers indicate that the client is being throttled.

## Note:

This is particularly important for managing Terraform operations using a Cluster DataSource to create a large number of Dataplane resources (topics, acls, users)

### PR DESCRIPTION
**Pull request by gene-redpanda**
https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-ratelimit-headers#name-ratelimit-header-field-defi

Relevant RFC used on the API side. I also added a bulk test case and info on what it's doing there. 

Fixes #108

### CODE REVIEW

File: redpanda/cloud/ratelimiter.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	limiter *rate.Limiter
+}
+
+func newRateLimiter(limit int) *rateLimiter {
+	return &rateLimiter{
+		limiter: rate.NewLimiter(rate.Limit(limit/limitPeriod), limit/burstPeriod),
+	}
+}
+
+func parseRateLimit(header string) (limit, remaining int, reset time.Duration, err error) {
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:31]: Please document what it does and if possible add a test for this function. Adding the expected format of the header to parse can help here

File: redpanda/cloud/ratelimiter.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	for _, part := range strings.Split(header, ",") {
+		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
+		if len(kv) != 2 {
+			continue
+		}
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:36]: What does it mean to have a length different from 2, is it a malformed ratelimit header and should be ignored?

File: redpanda/cloud/ratelimiter.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		switch kv[0] {
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:37]:
```suggestion
                keyword, val := kv[0], kv[1]
		switch keyword {
```

And then use the value in the error messages below

File: redpanda/cloud/ratelimiter.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		case "limit":
+			limit, err = strconv.Atoi(kv[1])
+			if err != nil {
+				return 0, 0, 0, fmt.Errorf("invalid limit value: %v", err)
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:41]: For example:
```suggestion
				return 0, 0, 0, fmt.Errorf("invalid limit value %q: %v", val, err)
```

File: redpanda/cloud/ratelimiter.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+			case <-time.After(reset + 1*time.Second):
+			}
+		}
+		tflog.Debug(ctx, "rate limit updated", map[string]any{
+			"limit":     limit,
+			"remaining": remaining,
+			"reset":     reset,
+		})
+	}
+	return r.limiter.Wait(ctx)
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:93]: Do we have to call `limiter.Wait` in every interceptor execution even if the header `ratelimit` is not set (L65)?

If we don't, we could:
```Go
if rateLimitHeader := header.Get("ratelimit"); len(rateLimitHeader) > 0 {
  ...
  r.limiter.Wait(ctx)
}
return nil
```

File: redpanda/cloud/ratelimiter.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	limitPeriod = 60.0
+	burstPeriod = 10.0
```
Comment (r-vasquez) [inline redpanda/cloud/ratelimiter.go:18]: please document these variables and why they are 60.0 and 10.0

File: Makefile [hunk 7]
```diff
@@ -64,5 +85,44 @@ linter:
...
+	@$(GOLANGCILINTCMD) run --config=.golangci.yml
+
+
+# Allow overriding these variables from the environment
+OS ?= $(shell go env GOOS)
+ARCH ?= $(shell go env GOARCH)
+PROVIDER_VERSION ?= 0.5.2
+PROVIDER_NAME ?= redpanda
+PROVIDER_NAMESPACE ?= redpanda-data
+TF_CONFIG_DIR ?= examples/bulk-res
```
[the remark this answers is not in the document]
Comment (gene-redpanda) [reply Makefile:97]: Intent was to have bulk examples for cluster creation with resource and just resource creation from a datasource cluster. I... didn't think of a better way to name the directories that wouldn't also bloat the directory name enormously LOL

File: redpanda/cloud/cloud.go [hunk 8]
```diff
@@ -112,6 +113,8 @@ func RequestTokenAndEnv(ctx context.Context, cloudEnv, clientID, clientSecret st
 	return tokenContainer.AccessToken, &endpoint, nil
 }
 
+var rl = newRateLimiter(500)
```
[the remark this answers is not in the document]
Comment (gene-redpanda) [reply redpanda/cloud/cloud.go:116]: The implementation is safe across multiple connections from my testing and from what I've read in the underlying package. 

My understanding is also that we need to rate limit at the client level not the individual connection level.

File: redpanda/cloud/ratelimiter.go (hunk 5 above)
Comment (gene-redpanda) [reply redpanda/cloud/ratelimiter.go:93]: Yes. But if nothing is set wait time is 0.

File: redpanda/cloud/ratelimiter.go (hunk 6 above)
Comment (gene-redpanda) [reply redpanda/cloud/ratelimiter.go:18]: Derived from the rate limit

File: redpanda/cloud/ratelimiter.go (hunk 1 above)
Comment (gene-redpanda) [reply redpanda/cloud/ratelimiter.go:31]: Test for the function is part of the PR? https://github.com/redpanda-data/terraform-provider-redpanda/blob/5500475701c29e28b7053ca8b6633eec82e05794/redpanda/cloud/ratelimiter_test.go

Comment (r-vasquez) [reply redpanda/cloud/ratelimiter.go:31]: The documentation helps for future maintainability and to understand what are we trying to parse.

btw, the test you linked is a test for the ratelimiter (which uses the `parseRateLimit`). However, it does test the underlying function

File: redpanda/cloud/ratelimiter.go (hunk 5 above)
Comment (r-vasquez) [reply redpanda/cloud/ratelimiter.go:93]: Ok, then we can avoid executing the function if we limit the usage to be within the block of code that actually will modify what it does, right?

File: redpanda/cloud/ratelimiter.go (hunk 2 above)
Comment (gene-redpanda) [reply redpanda/cloud/ratelimiter.go:36]: Essentially, yeah

**Review by r-vasquez — APPROVED** [review]
LGTM! Thanks

**Pull request merged**